### ISSUE TITLE
First steps

### ISSUE THREAD

**Issue by gabrielcsapo** [gabrielcsapo/groffee#2]
Just wanted to see where we can go from here @echopoint @plunkettscott @jlxip

**Comment by jlxip** [comment]
Are we going to store the data in a DB? Or how are we going to do it?

You can copy as much code from UltraGit as you want if it can help you.

**Comment by gabrielcsapo** [comment]
@jlxip I assume we are going to store information in a DB. Github has an interested way of having federated control at the application level https://enterprise.github.com/features which means they are constantly listening for data and synchronizing it between nodes. Interesting this can be done using node-git-server as seen here https://github.com/kubenstein/starbucket fairly easily.

**Comment by plunkettscott** [comment]
Github Enterprise does a lot of things well. How in-depth are you wanting to go with this? Are you wanting to allow searching the repositories like Github does?

Scott

**Comment by jlxip** [comment]
Github Enterprise? I was seeing this project as if no money were related. I don't think this is a startup.

**Comment by gabrielcsapo** [comment]
@plunkettscott I was just looking at examples of well established ideations of this, I think it should be designed with expansion in mind rather than creating it over and over again. 

@jlxip definitely not a startup, just using it as a reference project with well established direction and material.

**Comment by echopoint** [comment]
Keep it simple and extendable. Take a page from fossil's book. Up front we we just need the basics to manage the repos and user access, possibly issue tracking. Can't see much more than that up front.

**Comment by gabrielcsapo** [comment]
@echopoint I am up for that. Any technology decisions? sqlite? I want to keep the dependency tree as small as possible as to avoid any c++ dependencies that could impair the platform from being deployed on any device that could run node.

**Comment by echopoint** [comment]
Avoiding C/C++ dependencies kills almost all dbs unless they have a js implementation of their client protocol implemented. I would make the database piece a simple model that would need to be implemented no matter what the backend would be. Then we can just keep adding extensions for different db's, etc, via a model plugin that handles that db or even lower but there isn't really much need for much lower level than that interface. This way it's transparent to the ui and the rest of the system as to what db is being used in the model. Using such an interface you could also ditch the db altogether and store the records however you want and it wouldn't matter, even as json on the fs and it would still work appropriately.

**Comment by gabrielcsapo** [comment]
@echopoint that makes sense. SQLite would be a good initial DB implementation for the ORM. Mongo has a fully Javascript implementation library so that could be another choice as well.

**Comment by echopoint** [comment]
oh dear, please no orm, haha... just an expectation of the interface that faces the app is enough. orms are  a huge mess of franken spaghetti 😬

**Comment by gabrielcsapo** [comment]
@echopoint as I see! Okay haha no ORM 😂

**Comment by gabrielcsapo** [comment]
@echopoint thoughts on first steps? Should we come up with a list of features we want to see first?

**Comment by echopoint** [comment]
Yes, we should decide what's needed up front and strip that list down into tasks. 😄
So far we have :

- [ ] basic site design and layout
- [ ] user creation and management
- [ ] user repo management
- [ ] administration controls
- [x] app framework skeleton

**Comment by gabrielcsapo** [comment]
- [ ] repo meta data
   - [ ] issues
   - [ ] pages
   - [ ] access (private or public)
- [ ] search (this would depend on access control)
- [ ] backup (would need a data format to contain database information as well as repo data)

**Comment by gabrielcsapo** [comment]
@echopoint I will have some time this weekend to get an MVP done to have something to discuss more specifically about if that works?

**Comment by gabrielcsapo** [comment]
@echopoint for right now each view can be stateful via the url so for example 

> /:name/repos 

could be easily rendered using template strings 

```javascript 
function renderUserRepos(user) {
   const repos = await User.getRepos(user);
   return `
      <ul>
        ${repos.map((repo) => {
          return `<li> ${repo.name} </li>`
        }).join('\n')}
      </ul>
    `;
}
```

Again this is just really quick example of what I think can be done with no UI framework, but bringing react and such in would have added benefits for being able to test these views.

**Comment by echopoint** [comment]
I have a router in the app framework locally that can handle that kind of route just fine. So no worries. Let me finish the part I was working on and I'll push the bare bones up. I didn't create any form of template processing for it at all for views / renders. Though if you want to use react, then go for it.

```javascript
var server = App.create("https://localhost:7000");
server.get('/api/:ver/', async function(conn, ver) {
	conn.content = "Get request: " + conn.method + " :: " + conn.uri + "<br/>\n" 
				 + util.inspect(conn.query);
	return 200;
});
server.start();
```

**Comment by gabrielcsapo** [comment]
@echopoint depends on what you want to do, we can make it work and refine later. You rock, can't wait to see what you have!

**Comment by echopoint** [comment]
well I needed the fw and I don't exactly like express, but don't hate it either.. and for views / ui, I just didn't do anything because that always differs. depending on the project. I like working on the server side more anyway, though I have no issue with react. I opted for my own fw before just didn't know if we could use it here or not because not sure what everyone is comfortable with.

**Comment by echopoint** [comment]
btw, if we use this thing, I didn't allow regex in the routing. call me lazy. Though, I like cleaner and easier to read interfaces that are self descriptive more often than not. 🙈

**Comment by gabrielcsapo** [comment]
@echopoint sounds good to me! Certain routes we might need to have regex for or we might to dynamically generate the routes and their handlers. For example if we add a user `foo` then we could procedurally generate:

- `/foo/profile`
- `/foo/repo`
- `/foo/settings`

No need for regex at that point.

**Comment by echopoint** [comment]
- /:user/profile
- /:user/repo
- /:user/settings

This does the same thing, but captures the name in :user and sends it to the handler with the request.
i.e. /:user/profile would call handler(conn, 'foo') or handler(conn) with conn.params.user = 'foo'.. And if we need regex in the handler route uri, then I'll add it. just didn't want to deal with it at that point in time 🐙

**Comment by gabrielcsapo** [comment]
Yeah I totally understand, I am saying if we didn't want to implement regex, we could use dynamically created routes. 🦑

**Comment by gabrielcsapo** [comment]
@echopoint do you want ownership of this repo so we can move the trello kanban to github projects?

**Comment by echopoint** [comment]
No need.. That trello was for my reference 😃  If we use the fw, then I'll worry about that. Right now I'm just slowly reapplying pieces back to it that I was working on earlier and rechecking it.. Basically I made it because I was bored, but it may be a good idea.

**Comment by gabrielcsapo** [comment]
@echopoint I like it! The DX in the examples in `server.js` looks really nice. The reason I am not a fan of express is the fact that you can set priority in the middleware chain by default.  https://github.com/krakenjs/kraken-js fixes this (I work at PayPal on the NodeInfra team) and it is pleasant, but if you can do that without leveraging express I would be super happy to use that as a dependency.

**Comment by gabrielcsapo** [comment]
> side note, whatever we make, please let's make it nicer than bitbucket...Github is so simple compared to Bitbucket and Gitlab.

**Comment by echopoint** [comment]
Well, that's a matter of time. It's already that far without leveraging express 🤐

**Comment by gabrielcsapo** [comment]
@echopoint awesome!

**Comment by echopoint** [comment]
@gabrielcsapo modules are in, let me know if that fits for you on them.

**Comment by gabrielcsapo** [comment]
@echopoint I am good on anything, we can refine when everything gets in

**Comment by gabrielcsapo** [comment]
@echopoint do you want to open a PR or do you want me to?

**Comment by echopoint** [comment]
talking about the fw? I'll finish the last couple of changes I was in the middle of and push it up.

**Comment by gabrielcsapo** [comment]
@echopoint 🙌 awesome!

**Comment by echopoint** [comment]
We should start thinking about an org for our stuffs. Starting to get quite the collection 😆

**Comment by gabrielcsapo** [comment]
@echopoint we can definitely move it to an org, do you want to do that?

**Comment by echopoint** [comment]
sure. we need a name for it. 📌

**Comment by gabrielcsapo** [comment]
@echopoint true true what about 🚀https://www.npmjs.com/package/ghan which is a name of a famous train 🚂 that would be awesome! I don't know I am bad at names that took me 40 minutes

**Comment by echopoint** [comment]
Only reason I wouldn't jump on that is the association to an in service train. Don't know how much they will care, as it's meant to be an homage.  What about allegiant? That's the only alternative I have atm.

**Comment by gabrielcsapo** [comment]
who is that is a cool name!

**Comment by echopoint** [comment]
@gabrielcsapo ok, allegiant it is.. made and you're a co-owner 🌠 
I'll close the pr and add the fw as a package for allegiant.

**Comment by gabrielcsapo** [comment]
You rock! So excited for this man!

**Comment by echopoint** [comment]
@gabrielcsapo created an org on github to collect all repos in for npm.. I figured this was the better way: [allegiant-js](https://github.com/orgs/allegiant-js). You're also you're co owner, as soon as you accept. 👏

**Comment by gabrielcsapo** [comment]
@echopoint accepted, are you going to add the allegiant framework there?

**Comment by echopoint** [comment]
Yes and splitting out the libs appropriately.

**Comment by echopoint** [comment]
@gabrielcsapo core is up.. that's the most basic core module needed, hence core. Still things to do, but that can come with time, since full docs are still needed, and activating coverage on most of the other modules. If you need to make a change, and the build passes travis feel free to commit. I've been under the weather the last 2 days, so taking the rest of the night off and tomorrow mostly off.

**Comment by gabrielcsapo** [comment]
I have also been DOA @echopoint I am enjoying some time off before I start a new job so anything new is slowly but surely making it on my todo list. If you need help with core, I can start closing out issues if you open them up!

**Comment by echopoint** [comment]
Already fixed the issues that cropped up from splitting the modules out as they were being posted. So far it's working fine. Since I can switch between these and my local original, without any issues at all, then it's definitely right.

**Comment by gabrielcsapo** [comment]
Miss everyone, hope you are all well, thoughts on making this a thing again?

**Comment by echopoint** [comment]
no problems from this end, just need to define the need and go from there

**Comment by plunkettscott** [comment]
^^

**Comment by gabrielcsapo** [comment]
I think the biggest need from my standpoint is referential material in how git works and how this can be used as educational material by leveraging its patterns to build something similar to github, but students or educators can use to build new features into very easily. I sometimes find myself asking wouldn't it be cool if Github was just one of many git server implementations that people could help maintain and add features to. If they didn't like the course they could use their knowledge to start a new community or even join another one that has a more aligned course (direction) with theirs.

**Comment by vadolasi** [comment]
Hello, is there any change for this project to actually start?

**Comment by gabrielcsapo** [comment]
Hello! Yeah I would like to work on it! Just don't know when I will have time. I might be more free during the holidays.

**Comment by gabrielcsapo** [comment]
after 9 years I can say this project has actually started, what a crazy amount of time.
